### DISCUSSION TITLE
Is it possible to insert userscripts directly into Webview-based app's data directory?  (root based)

### DISCUSSION THREAD

**Discussion by directentis1** [JingMatrix/ChromeXt#163] [General]
I have some Webview-based applications, and I can successfully insert the Eruda console into that application via ChromeXt hooks.

 Because there is no way to navigate to the `user.js` files from the outside, I tried to find a way to load `user.js` files from the Eruda console but it also failed because the application did not declare access rights to the filesystem so it could not call file explorer.

 Is there a way to load Userscripts directly into the application's Webview data directory without having to go through the Web interface?

 I noticed that in the App's data - webview dir folder, `Eruda.js` is downloaded in the `"files/"` folder, while for Chrome, the userscripts are loaded into the `"databases/"` folder with two files `"userscript"` and `"userscript -journal"` with sqlite format.

**Comment by JingMatrix** [comment]
Currently, userscripts are stored in sqlite database, i.e., the `databases/userscript` file.
You may directly change it, if you feel confident.

Could you open the ChromeXt front-end via Eruda? I may change front-end so that you can jump to valid userscript urls.
Without filesystem permission, it is really quite awkward.

**Comment by directentis1** [comment]
@JingMatrix I guess I'll need to make a copy of `/databases/userscript` files containing the `user.js` files I want before editing and move them to the destination folder.

 But not sure if it can load when opened in the other webview app.

 > Could you open the ChromeXt front-end via Eruda?  I may change the front-end so that you can jump to valid userscript urls.

 Can you guide me how to do that?  I'm not sure if there is any option to open ChromeXt front-end with Eruda.
 I just want to be able to insert some `user.js` files to process text, or translate text on some magazine/news apps that have webviews in them.

 > Without filesystem permission, it is really quite awkward.

Yes, so I think the only way is to move that userscripts file directly to the app's data folder (internal storage or sdcard) because the application has access to the filesytem granted to themselves
 Or the way you mentioned is to embed the url of `user.js` directly into Eruda, but not sure where it will be downloaded and how it will be processed afterwards.

**Reply by directentis1** [reply]
My phone's screen record.

https://github.com/JingMatrix/ChromeXt/assets/128832114/f67b99c6-01be-403c-95a1-d69dc3316cb7

**Comment by JingMatrix** [comment]
To open the front-end, try what I said in this link: https://github.com/JingMatrix/ChromeXt/issues/161#issuecomment-2083643869

Your scripts can definitely load once you edit the sqlite file properly.

My suggestion was that in the front-end, I would made an input box, through which you can jump to any url; this will then allows to intall scripts.

**Reply by directentis1** [reply]
> If you can open Eruda console, then you can open the manager page, and click Info -> UserScripts -> the eye icon.

Ah, I've tried that before, it'll fire an intent to open external browsers.
Well, so I wish I could add the url directly from the input box - like with CSP rules, that would definitely be better. 😄


https://github.com/JingMatrix/ChromeXt/assets/128832114/d5c2a73a-9fb1-4442-9b66-1ef60ceb1845

**Reply by directentis1** [reply]
I've tried to open userscript database, but there's nothing special (and there are no specific data fields, strings related to my target application)

Try to move userscript database to the `databases/` in app's data, because at first I forgot to change the ownerships back to the file (still be root), and the Eruda console could not be loaded inside the application's WebView like before;  change the permissions after that and it loads as usual but still don't see the scripts I added in Eruda's Info Tab.

![](https://i.imgur.com/bjD9Vsk.png)

**Reply by JingMatrix** [reply]
Well, I have a quick solution for you!

You install a browser, say Chrome, in which you can installed scripts without any restrictions.
And then you copy the `userscript` database file from the browser app folder to your WebView application browser.
Of course, respect the folder path.

**Reply by directentis1** [reply]
Yes, I did that already, but for some reason the script file still doesn't load when I open the Eruda console....

**Reply by JingMatrix** [reply]
Well, here are few things to check:
1. Could you upload a screenshot of your database? Note that you should click on the `Browser Data` tab in `sqlbrowser`.
2. Did the script worked as expected in your browser (Chrome)?
3. Did you fully restart your applications? You may try reboot your phone to ensure this.

**Reply by directentis1** [reply]
1. Well, I just use Immersive Translate js as a test.
So, it's like this:

![](https://i.imgur.com/PhWMIEN.png)
![](https://i.imgur.com/WOq2E2x.png)

The first 3 data fields are nothing worth discussing, but the 4th field, I checked its json file (about 8k lines, after beautifying it), but there's really nothing specially.

2. Yes, it works perfectly fine with Chrome.
3. I cleared the cache, closed the application, restarted,... but strangely it's still the same.

**Reply by JingMatrix** [reply]
It could be a problem of file permission.
1. Ensure the datebase is put in the correct folder;
2. Change the owner or permissions so that the app could access to it.
